db/constraints.c

```diff
@@ -2141,97 +2141,160 @@ static inline int key_has_expressions_members(struct schema *key)
+ *
+ * Logs each failure and returns the number of failed verifications.
+ */
+static int verify_constraint(constraint_t * const ct,
+                             struct dbtable *source_db,
+                             struct dbtable *target_db,
+                             int source_db_is_updated,
+                             int target_db_is_updated,
+                             struct schema_change_type *s)
+{
+    int n_errors = 0;
+
+    char keytag[MAXTAGLEN];
+    if (source_db_is_updated) {
+        snprintf(keytag, sizeof(keytag), ".NEW.%s", ct->lclkeyname);
+    } else {
+        snprintf(keytag, sizeof(keytag), "%s", ct->lclkeyname);
+    }
+
+    struct schema *fky = find_tag_schema(source_db, keytag);
+    if (!fky) {
+        constraint_err(s, source_db, ct, 0, "foreign key not found");
+        n_errors++;
+    }
+    if (source_db->ix_expr && key_has_expressions_members(fky) &&
+        (ct->flags & CT_UPD_CASCADE)) {
+        constraint_err(s, source_db, ct, 0, "cascading update on expression indexes is not allowed");
+        n_errors++;
+    }
+    for (int jj = 0; jj < ct->nrules; jj++) {
+        const int constraint_rule_is_eligible_for_verification = !target_db
+            || strcasecmp(ct->table[jj], target_db->tablename) == 0;
+        if (constraint_rule_is_eligible_for_verification) {
+            n_errors += verify_constraint_rule(ct, jj, source_db, target_db_is_updated ? target_db : NULL, fky, s);
         }
     }
 
     return n_errors;
 }
 
+int verify_constraints_from_dbtable(struct dbtable *source_db,
+                                    int is_updated,
+                                    struct schema_change_type *s)
+{
+    int n_errors = 0;
+
+    for (int ii = 0; ii < source_db->n_constraints; ii++) {
+        constraint_t * const ct = &source_db->constraints[ii];
+        n_errors += verify_constraint(ct, source_db, NULL,
+            /* src_db_is_updated */ is_updated, /* target_db_is_updated */ 0, s);
+    }
+
+    return n_errors;
+}
+
+static int verify_constraints_to_dbtable(struct dbtable *target_db,
+                                         int is_updated,
+                                         struct schema_change_type *s)
+{
+    int n_errors = 0;
+    for (int ii = 0; ii < thedb->num_dbs; ii++) {
+        // Need to call `get_newer_db` for the case when there is a self-referencing constraint.
+        struct dbtable * const source_db = get_newer_db(thedb->dbs[ii], target_db);
+
+        for (int ii = 0; ii < source_db->n_constraints; ii++) {
+            constraint_t * const ct = &source_db->constraints[ii];
+            n_errors += verify_constraint(ct, source_db, target_db,
+                /* source_db_is_updated */ 0, /* target_db_is_updated */ is_updated, s);
+        }
+    }
+    return n_errors;
+}
+
+int verify_constraints_to_and_from_dbtable(struct dbtable *new_db, int is_updated, struct schema_change_type *s)
```

Comment:
Considering renaming all `is_updated` variables to `undergoing_rebuild`/`is_rebuild`/`is_rebuild_needed`